e2e/bitrise.yml

```diff
@@ -0,0 +1,688 @@
+            - pull_request_repository_url: $TEST_DIFF_PR_REPOSITORY_URL
+            - pull_request_merge_branch: ""
+            - branch_dest: "master"
+            - branch: "master"
+            - commit: "c6810e6"
+            - clone_depth: "1"
+            - manual_merge: "no"
+      - script:
+          inputs:
+            - content: |-
+                #!/bin/env bash
+                set -ex
+                diff ${TEST_DIFF_CLONE_DIR}/README.md ${TEST_DIFF_PR_REPOSITORY_URL}/README.md
+    after_run:
+      - _log_outputs
+
+  test_sparse_checkout_branch:
+    before_run:
+      - _activate_ssh_key
+      - _prepare
+    steps:
+      - path::./:
+          run_if: true
+          inputs:
+            - repository_url: $TEST_REPO_URL
+            - clone_into_dir: $WORKDIR
+            - branch: "master"
+            - sparse_directories: |-
+                git-clone-sample
+    after_run:
+      - _log_outputs
+      - _ensure_clean_git
+
+  test_sparse_checkout_commit_on_branch:
+    before_run:
+      - _activate_ssh_key
+      - _prepare
+    steps:
+      - path::./:
+          run_if: true
+          inputs:
+            - repository_url: $TEST_REPO_URL
+            - clone_into_dir: $WORKDIR
+            - commit: 7a93251a9bf08fb8c6b19fc02f6d94618de6b386
+            - branch: "master"
+            - sparse_directories: |-
+                git-clone-sample
+    after_run:
+      - _log_outputs
+      - _ensure_clean_git
+
+  test_sparse_checkout_unshallow:
+    before_run:
+      - _activate_ssh_key
+      - _prepare
+    steps:
+      - path::./:
+          run_if: true
+          inputs:
+            - repository_url: $TEST_REPO_URL
+            - clone_into_dir: $WORKDIR
+            - commit: cfba2b01332e31cb1568dbf3f22edce063118bae
+            - clone_depth: 1
+            - sparse_directories: |-
+                git-clone-sample
+    after_run:
+      - _log_outputs
+      - _ensure_clean_git
+
+  test_too_long_commit_message:
+    before_run:
+      - _prepare
+    steps:
+      - path::./:
+          run_if: true
+          inputs:
+            - repository_url: $TEST_REPO_URL
+            - clone_into_dir: $WORKDIR
+            - commit: "2e8cb1e06cda054e6eb861e56761acd52441cd86"
+            - update_submodules: "no"
+      - script:
+          inputs:
+            - content: |-
+                #!/bin/env bash
+                # --- This test only works for default 20kbyte env sizes, if altered, test should be modified accordingly.
+                if [ ${#GIT_CLONE_COMMIT_MESSAGE_BODY} -gt "20480" ]; then
+                  echo "Message body is bigger than 20480 chars"
+                  exit 1
+                fi
+                if [ ${#GIT_CLONE_COMMIT_MESSAGE_SUBJECT} -gt "20480" ]; then
+                  echo "Message subject is bigger than 20480 chars"
+                  exit 1
+                fi
+    after_run:
+      - _log_outputs
+      - _ensure_clean_git
+
+  _prepare:
```

Comment:
Probably we should have a naming convention for the `setup` kind of workflows too like `prepare` now. We might also have `teardown` kind of wfs like `_ensure_clean_git` now

I'm thinking like

```
test_*                  Test workflows
utility_*               Workflows that are used from tests
_check_*             Verify or print results
_setup_*             Prepare state for test
_teardown_*      Clean up state after test
```

Reply:
You're right. Then I think we don't need the _setup/_teardown wfs just the _run but we will have to refactor this yml to that.

```
test_*                  Test workflows (prepare env vars for _run basically)
utility_*               Workflows that are used from tests
_check_*             Verify or print results
_run                    3 phase of step execution: setup, run the step, teardown
```